src/wp-includes/blocks.php

```diff
@@ -1073,31 +1086,57 @@ function update_ignored_hooked_blocks_postmeta( $post ) {
 	}
 
 	/*
-	 * Skip meta generation when the post content is not a navigation block.
+	 * We only hook into filters for the `wp_navigation` and `wp_template_part` post types.
+	 * So if the post type is not set we can assume it's a `wp_template_part`.
+	 */
+	$post_type                   = isset( $post->post_type ) ? $post->post_type : 'wp_template_part';
+	$post_type_to_block_name_map = array(
+		'wp_navigation'    => 'core/navigation',
+		'wp_template_part' => 'core/template-part',
+	);
+
+	/*
+	 * Skip meta generation when the post content is not in the above map.
 	 */
-	if ( ! isset( $post->post_type ) || 'wp_navigation' !== $post->post_type ) {
+	if ( ! isset( $post_type_to_block_name_map[ $post_type ] ) ) {
 		return $post;
 	}
 
-	$attributes = array();
+	$parent_block_name              = $post_type_to_block_name_map[ $post_type ];
+	$attributes                     = array();
+	$existing_ignored_hooked_blocks = isset( $post->ID ) ? get_post_meta( $post->ID, '_wp_ignored_hooked_blocks', true ) : '';
 
-	$ignored_hooked_blocks = get_post_meta( $post->ID, '_wp_ignored_hooked_blocks', true );
-	if ( ! empty( $ignored_hooked_blocks ) ) {
-		$ignored_hooked_blocks  = json_decode( $ignored_hooked_blocks, true );
+	if ( ! empty( $existing_ignored_hooked_blocks ) ) {
 		$attributes['metadata'] = array(
-			'ignoredHookedBlocks' => $ignored_hooked_blocks,
+			'ignoredHookedBlocks' => json_decode( $existing_ignored_hooked_blocks, true ),
 		);
 	}
 
 	$markup = get_comment_delimited_block_content(
-		'core/navigation',
+		$parent_block_name,
 		$attributes,
 		$post->post_content
 	);
 
-	$existing_post = get_post( $post->ID );
-	// Merge the existing post object with the updated post object to pass to the block hooks algorithm for context.
-	$context          = (object) array_merge( (array) $existing_post, (array) $post );
+	/**
+	 * We need to merge incoming changes with existing data so the block hooks API gets the correct context.
+	 * For `wp_template_part` post type, we need to convert the post object to a block template object.
+	 */
+	$existing_post = isset( $post->ID ) ? get_post( $post->ID ) : array();
+	$context       = (object) array_merge( (array) $existing_post, (array) $post );
+	// Given a post object ($context), get_post will return a WP_Post object.
+	$context = get_post( $context );
+	if ( 'wp_template_part' === $post_type ) {
+		// Convert the $context WP_Post object to a WP_Block_Template object.
+		if ( isset( $post->ID ) ) {
+			$context = _build_block_template_result_from_post( $context );
+		} else {
+			$meta    = isset( $context->meta_input ) ? $context->meta_input : array();
+			$terms   = isset( $context->tax_input ) ? $context->tax_input : array();
+			$context = _build_block_template_object_from_post_object( get_post( $context ), $terms, $meta );
+		}
+	}
+
```

Comment:
This code seemed fairly reminiscent of `inject_ignored_hooked_blocks_metadata_attributes`, which makes sense, given [this other observation](https://github.com/WordPress/wordpress-develop/pull/6867/files#r1660880711) 😅